File: 2020/38.py
class Solution:
    def countAndSay(self, n: int) -> str:
        res = (1,'1')       # res[i] is the nth-seq to 'say'
        if n == 0: return

        def say(x: str) -> str:
            output, cnt, i = "", 0, 0
            while i < len(x):
                cnt += 1
                if i+1 == len(x) or x[i] != x[i+1]:
                    output += str(cnt)+x[i]
                    cnt = 0     # it'll be fine to reset the counter when i+1=len(x)
                i += 1
            return output

        while res[0] < n:
            res = (res[0]+1, say(res[1]))
        return res[-1]

print(Solution().countAndSay(1))
print(Solution().countAndSay(4))
print(Solution().countAndSay(5))
print(Solution().countAndSay(10))
# print(Solution().countAndSay(4))

File: current_session/c++/38.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
  string countAndSay(int n) {
    if (n == 1) return "1";
    return helper(countAndSay(n-1));
  }
private:
  string helper(string s) {
    string res;
    int prevI = 0;
    int cnt = 1;
    for (int i=1; i<s.size(); ++i) {
      if (s[i] == s[prevI]) {
        cnt += 1;
      } else {
        res = res + to_string(cnt) + s[prevI];
        cnt = 1;
        prevI = i;
      }
    }
    res = res + to_string(cnt) + s[prevI];
    return res;
  }

};